connectors/migrations/20241211_fix_gdrive_parents.ts

```diff
@@ -0,0 +1,256 @@
+import logger from "@connectors/logger/logger";
+import { ConnectorModel } from "@connectors/resources/storage/models/connector_model";
+
+const QUERY_BATCH_SIZE = 1024;
+
+async function migrate({
+  connector,
+  execute,
+}: {
+  connector: ConnectorModel;
+  execute: boolean;
+}) {
+  logger.info(`Processing connector ${connector.id}...`);
+
+  const dataSourceConfig = dataSourceConfigFromConnector(connector);
+  const parentsMap: Record<string, string | null> = {};
+  let nextId = 0;
+  for (;;) {
+    const googleDriveFiles = await GoogleDriveFiles.findAll({
+      where: {
+        connectorId: connector.id,
+        id: {
+          [Op.gt]: nextId,
+        },
+      },
+    });
+
+    const last = googleDriveFiles[googleDriveFiles.length - 1];
+    if (!last) {
+      break;
+    }
+
+    googleDriveFiles.forEach((folder) => {
+      parentsMap[folder.driveFileId] = folder.parentId;
+    });
+
+    nextId = last.id;
+  }
+
+  nextId = 0;
+  for (;;) {
+    const googleDriveFiles = await GoogleDriveFiles.findAll({
+      where: {
+        connectorId: connector.id,
+        id: {
+          [Op.gt]: nextId,
+        },
+      },
+      order: [["id", "ASC"]],
+      limit: QUERY_BATCH_SIZE,
+    });
+
+    const last = googleDriveFiles[googleDriveFiles.length - 1];
+    if (!last) {
+      break;
+    }
+
+    for (const file of googleDriveFiles) {
+      const internalId = file.dustFileId;
+      const driveFileId = file.driveFileId;
+      const parents = [driveFileId];
+      let current: string | null = file.parentId;
+      while (current) {
+        parents.push(current);
+        current = parentsMap[current] || null;
+      }
+
+      if (file.mimeType === "application/vnd.google-apps.folder") {
+        const folder = await getFolderNode({
+          dataSourceConfig,
+          folderId: internalId,
+        });
+        if (folder) {
+          logger.info(
+            {
+              folderId: file.driveFileId,
+            },
+            "Found folder"
+          );
+          if (folder.parents.join("/") !== parents.join("/")) {
+            logger.info(
+              { folderId: file.driveFileId, parents },
+              "Upsert folder"
+            );
+
+            if (execute) {
+              // upsert repository as folder
+              await upsertFolderNode({
+                dataSourceConfig,
+                folderId: file.dustFileId,
+                parents,
+                parentId: file.parentId,
+                title: file.name,
+              });
+            }
+          }
+        }
+      } else {
+        const document = await getDocumentFromDataSource({
+          dataSourceConfig,
+          documentId: internalId,
+        });
+        if (document) {
+          logger.info(
+            {
+              documentId: file.driveFileId,
+            },
+            "Found document"
+          );
+          if (document.parents.join("/") !== parents.join("/")) {
+            logger.info(
```

Comment:
nit: we'll likely have ~10M lines of log for the run; maybe output in file? (that will I guess weigh in the ballpark of gigabyte(s))

Reply:
yes, log is already pretty long, will be better on one line

Reply:
some logs are generated by data_source and this does not allow to use a custom logger . it may be easier to redirect the script output to a file

Reply:
yes that's what I meant sorry for being unclear